### ISSUE TITLE
[News][Taxonomy][DSPy][Policy][P1] Current-epoch 联合质量闭环与 decide v12 受控重校准

### ISSUE THREAD

**Issue by aaurix** [AnalyThothAI/tracefold#399]
## 审计结论

`#117` 已完成 `news_taxonomy_v1` 的结构、typed output、Review/Evaluator substrate 与生产 EventSemantics 接入；`#369` 已删除在线 `event_type` 双轨并建立 current-only contract。但是两者都没有证明：

```text
taxonomy quality = PASS
TradeRelevance quality = PASS
current decide policy is calibrated against the new semantics
```

当前 `decide()` 刻意不把 taxonomy 当 admission authority，而是读取 `TradeRelevance.reader_value`、magnitude、tradability、channels/markets、novelty/development 以及 stale/duplicate/watchlist/listing 等确定性事实。这个边界在 taxonomy 质量仍为 `UNKNOWN` 时是正确的。

现在可以继续优化整体链路，但不能直接写：

```python
if taxonomy.event_family in {...}:
    push()
```

`event_family` 描述“发生了什么”，不是“是否值得推送”；`source_authority` 描述来源，不能自动证明交易价值；`assertion_status=rumor` 也不能等价为无市场影响。正确路径是：**先在 current epoch 证明 taxonomy/TradeRelevance，再以同一份 frozen Judgment 做离线 policy calibration，最后 shadow/canary/human promotion。**

## 目标

在不增加生产模型调用、不引入规则引擎、不建立第二套自研学习平台的前提下，完成：

1. current-epoch taxonomy 与 TradeRelevance accepted Gold；
2. 一个对 EventSemantics 有真实 field-level feedback 的 DSPy/GEPA 优化闭环；
3. current `decide v11` 与候选 `decide v12` 的同输入、可重放离线对比；
4. taxonomy 只在证据支持后进入 code-owned policy feature，而不是直接成为模型动作；
5. future holdout、shadow、10% canary、人工 promote/reject 与 rollback；
6. 最终分别声明分类、相关性和 policy 的 PASS/FAIL/UNKNOWN。

### 生产 Program 仍然只有两个 Predictor

```text
EventSemantics  -> typed facts + taxonomy + TradeRelevance
ReaderCard      -> headline_zh + why_zh
```

- 不增加第三个在线 taxonomy classifier。
- 不把 offline taxonomy shadow module 接入热路径。
- 保持 `dspy.Predict` + Pydantic/Literal exact output；分类默认不引入显式 ChainOfThought。
- `source_authority` 始终由代码从 exact structured provenance 计算，模型和 GEPA 均无写权限。
- ReaderCard 不拥有事实分类和 push/drop 权威。

### `decide()` 始终是确定性最终动作权威

DSPy 负责生成可审计语义事实；code-owned policy 负责最终：

```text
push / escalate / drop / throttled
```

模型不能输出最终动作。taxonomy 在第一阶段只能用于：

- evaluation slice；
- Review 采样；
- error attribution；
- semantic consistency diagnosis；
- policy candidate 的预注册分析。

任何 taxonomy-derived policy feature 都必须经过本 Issue 的离线对比、future holdout、shadow、canary 和人工 promotion，且先作为约束/校准证据，不能未经证明成为 positive push override。

## Phase 0 — 前置阻断必须先关闭

本 Issue 不复制 `#367` 已定义的缺陷，开始 provider-backed 优化或 policy canary 前必须完成：

- [ ] exact provenance parser；任意 `fake|sec`、`fan:reuters`、path fragment 不能提升 `source_authority`；
- [ ] execution identity 覆盖实际 routing/fallback/normalization/error semantics；
- [ ] taxonomy schema-invalid/parse-invalid 来自真实调用与真实分母，不能 hardcode `0/PASS`；
- [ ] routing 只捕获可归因 provider/transport failure，代码 bug/取消/进程异常不得伪装成 fallback；
- [ ] shadow recording 支持 adapter 的合法多次 physical call，并可严格 record/replay；
- [ ] `#398` current-contract genesis 完成，所有 Gold/holdout/shadow 分母从新 epoch 开始。

任何一项缺失均为 `UNKNOWN/BLOCKED`，不能解释性豁免。

## Phase 1 — current-epoch accepted Gold

复用 `#245` 和现有 `news_review_v6`/artifact ledger，不新建平行 corpus framework。

一个 release-eligible case 至少绑定：

- current Event/Evidence identity；
- fact cluster identity；
- exact taxonomy 五轴；
- exact TradeRelevance 七轴；
- assets/direction/magnitude/novelty 的 accepted value；
- should-push/should-hold judgement 及理由；
- reviewer/adjudicator/acceptance identity；
- Program/policy/model/runtime/schema identity。

同一事实的多来源重复只计一个 primary cluster；train、development-selection、future holdout 不得共享 cluster。critical family、模型/草稿分歧和来源状态分歧必须独立 adjudication。

采样不能只来自当前 push 流量，必须覆盖：

- pushed / held / dropped / degraded；
- product、financial、macro、market access、regulatory/legal、incident、flow/price、other/abstention；
- 中文/英文、多来源、first-party/filing/secondary/unknown；
- positive action、negative action、duplicate、stale、no-asset、wrong-asset；
- `other/unknown/empty subject_codes` 的真实 abstention 原因。

样本不足的 class/slice 必须显示 `UNKNOWN`，不能通过 macro average 隐藏长尾。Development/holdout 数量与 family/source/language coverage 复用 `#117/#245` 已冻结的 readiness contract，不在本 Issue 再造第二套分母。

## Phase 2 — DSPy EventSemantics 联合优化

当前通用 GEPA objective 把 taxonomy 排除在 reflective target 外，但 EventSemantics instruction 同时生成 taxonomy、TradeRelevance、asset、direction、magnitude 和 novelty。继续只优化部分输出，会允许候选在 relevance 得分提高时悄然破坏 taxonomy。

### 统一 metric，但不统一成一个黑盒分数

在现有 CandidateEvaluator/compile/release machinery 中，为 EventSemantics 返回 predictor-level feedback：

```text
taxonomy_subject
taxonomy_event_family
taxonomy_change_state
taxonomy_assertion_status
trade_relevance
asset_grounding
direction
magnitude
novelty
production_action_retention
schema/parse validity
```

使用 GEPA named objective scores / textual feedback：

- 每个轴独立 score 与 failure message；
- 指出 expected/observed、冲突规则、evidence span 与 owner；
- schema/parse/constraint failure 为硬失败；
- taxonomy accepted Gold 可以成为 EventSemantics target/control；没有 exact Gold 的 case 继续 excluded；
- ReaderCard feedback 只进入 ReaderCard predictor，不污染 EventSemantics；
- production action 只作为候选非回归结果，不作为模型可直接模仿的动作标签。

不得再开发一套“taxonomy optimizer platform”；复用当前 Program artifact、candidate、evaluation、shadow、canary、promotion 和 receipt。

### 优化配置

- trainset 用于 reflective mutation；development-selection 用于候选选择；future holdout 完全隔离。
- reflection LM 使用独立、能力更强的 endpoint，不与 production task LM/GPU slot 共用。
- 先做 prompt/program optimization；只有在 prompt 已稳定、accepted Gold 数量与长尾覆盖足够、且 held-out plateau 可复现后，才另立 finetune proposal。
- 不因分类任务盲目启用 ChainOfThought；需要 reasoning 的收益必须由独立实验和 latency/token regression 证明。
- 候选必须保存 `detailed_results`、per-objective score、失败 cluster 和 exact instruction diff。

## Phase 3 — `DecisionCalibrationReportV1`

建立一个薄的、纯离线、可重放报告；它不是新 policy engine。

对每个 frozen accepted Judgment，同时运行：

```text
baseline:  decide v11 + exact frozen policy context
candidate: decide v12 + exact same Judgment/context
```

报告至少包含：

- final action confusion matrix；
- must-push retention / must-hold leakage；
- candidate-only regressions；
- push/escalate/drop/throttled flip list；
- duplicate、stale、listing、watchlist、no-asset 等 guard 行为；
- taxonomy 五轴 × TradeRelevance 七轴的 slice；
- source/language/audience/family/state/assertion slice；
- abstention coverage 与 error rate；
- baseline/candidate policy identity 与 dataset root。

### v12 变更规则

- 一次 candidate 只改变一个明确的 code-owned rule 或一个可解释 interaction；禁止一次性权重矩阵/DSL/通用规则引擎。
- 所有阈值和预期方向在运行 holdout 前 preregister。
- taxonomy 默认只用于 slice/consistency evidence；要进入 policy，必须展示它解决的具体 accepted failure 和不会制造的相邻 regression。
- 禁止 `family -> push`、`authority -> push`、`rumor -> drop` 这类单轴捷径。
- 不允许 GEPA 修改 `decide()`；GEPA 优化 Program，policy candidate 由代码审查和 sealed report 管理。

## Phase 4 — 质量门槛

延用 research、`#117` 与现有 evaluator 的预注册阈值：

- [ ] event-family macro-F1 `>= 0.85`；
- [ ] product_service_change precision 与 recall 各 `>= 0.90`；
- [ ] financial_results + guidance_outlook precision 与 recall 各 `>= 0.90`；
- [ ] change-state exact accuracy `>= 0.90`；
- [ ] known structured provenance 上 source-authority exact accuracy `= 1.00`，不确定时必须输出 `unknown`；
- [ ] assertion-status macro-F1 `>= 0.90`；
- [ ] subject-codes micro-F1 `>= 0.85`；
- [ ] non-abstain coverage `>= 0.80` 且 error rate `<= 0.08`；
- [ ] schema/cardinality/code invalid `= 0`，且来自真实 denominator；
- [ ] provider duplicates cluster-collapse 正确；
- [ ] train/dev/future-holdout cluster leakage `= 0`；
- [ ] production action、asset grounding、novelty、TradeRelevance 四类 candidate-only regression `= 0`；
- [ ] must-push/must-hold、long-tail slice 的其余阈值在 holdout 前封存；
- [ ] 分母不足一律 `UNKNOWN`，不得用空集 PASS。

这些阈值只证明分类/语义质量；`decide v12` 是否更好还必须由独立 DecisionCalibration gate 决定。

### Shadow

- 对同一个 persisted Judgment 同时计算 v11/v12 action，不增加模型调用；
- v11 仍为唯一 reader authority；
- 至少 `>=24h`、`>=200` eligible current-epoch Events、`>=30` accepted primary clusters；
- 输出 action flip、regression、latency、failure 与 per-slice report；
- 任何 identity mismatch、invalid denominator 或 missing Gold 均阻止进入 canary。

### Canary

- deterministic 10% eligibility，stable/candidate 使用同一 current Program Judgment；
- candidate-only must-hold leakage、duplicate/stale regression、delivery error 或 schema invalid 立即 trip；
- canary receipt 绑定 exact code/image/runtime/program/policy/dataset identities；
- 只有人工 reviewed diff + promote/reject 才能切换 stable；不得自动 promotion。

### Rollback

- policy rollback 只回到上一枚 stable policy artifact；
- 不恢复 event_type、旧数据库、兼容字段或 archive reader；
- rollback drill 必须证明 current-contract Event/Judgment 可继续处理。

## 与 research 的一致性

本 Issue 延续 `docs/research/news-classification-audit-and-survey-2026-08-29.md` 的核心顺序：

```text
正交 taxonomy
-> accepted Gold
-> cluster-disjoint evaluation
-> release-neutral shadow
-> controlled policy/canary
-> human promotion
```

actor/entity registry、geo episode 和 post-delivery Deep Analysis 是 taxonomy 的下游能力，不应继续塞入 `event_family`，也不应与本次 `decide` calibration 一起实现。

DSPy 实施参考：

- https://dspy.ai/getting-started/gepa-optimization/
- https://dspy.ai/api/optimizers/GEPA/overview/
- https://dspy.ai/tutorials/classification_finetuning/

### Program / DSPy

- [ ] production success path 仍恰好两个 Predictor。
- [ ] EventSemantics typed schema 不新增最终动作字段。
- [ ] GEPA metric 返回 per-predictor、per-objective score + actionable feedback。
- [ ] train/dev/future holdout 完全 cluster-disjoint。
- [ ] reflection endpoint 与 task endpoint/runtime slot 分离。
- [ ] optimized artifact 可保存、record/replay、diff 和按 exact identity 重跑。

### Classification / relevance

- [ ] current-epoch accepted Gold 达到 readiness；所有预注册 taxonomy gate 为 PASS。
- [ ] TradeRelevance、asset、novelty、direction、magnitude 无 candidate-only regression。
- [ ] schema-invalid/parse-invalid 使用真实 call denominator。
- [ ] source authority exact provenance tests 覆盖伪造 delimiter/domain/path。

### Policy

- [ ] `DecisionCalibrationReportV1` 对 v11/v12 使用完全相同 Judgment/context。
- [ ] v12 仅含一个可解释、预注册的 policy delta。
- [ ] 没有 raw taxonomy single-axis push/drop shortcut。
- [ ] must-push retention、must-hold leakage、duplicate/stale 与 action-flip gate 全部达到预注册阈值。

### Release

- [ ] future holdout、`>=24h` shadow、10% canary、人工 promote/reject、rollback drill 完成。
- [ ] exact-main CI、部署镜像、runtime manifest、Program/policy/dataset/evaluator identity 一致。
- [ ] terminal receipt 分别声明 taxonomy、TradeRelevance、policy 的 PASS/FAIL；任何缺失均为 UNKNOWN。

## 非目标

- 不新增在线第三分类器、agent 或 retrieval。
- 不把 taxonomy 直接接到 Trading/Capital。
- 不在本 Issue 做 actor registry、geo episode、Deep Analysis。
- 不引入通用 policy DSL、可配置权重森林或自研 AutoML。
- 不在 current Gold 尚少时启动生产 finetune。
- 不保留旧 event_type/legacy data compatibility；数据库清场由 `#398` 完成。

## 依赖与复用

- `#367`：先关闭 provenance、identity、invalid-output、exception、shadow replay 缺陷。
- `#398`：先建立空的 current-contract evidence epoch。
- `#245`：复用 accepted Gold、baseline、candidate evaluator 与 release evidence；不复制 corpus machinery。
- `#369`：复用 current-only contract 和 terminal holdout/canary barrier。
- `#344/#199`：复用 native DSPy 与既有 learning simplification，不恢复手写 Prompt/optimizer 双轨。

**Comment by aaurix** [comment]
## 范围裁决 — 2026-08-31

#437 接管 taxonomy Gold admission 与 Stable 分类测量；它取代 #399 针对该测量定义的分类 Gold/readiness floors 及 taxonomy shadow/canary 前置条件。

#399 只保留为未来 policy calibration 范围；在依据 #437 的真实结果重写之前，不得实施。#437 的 `MEASURED` 不是 policy release evidence，也不授权 `decide v12`、Prompt candidate、shadow、canary、promotion 或任何 Gate/Delivery 行为变更。

**Comment by aaurix** [comment]
## 最终处置：预设式联合校准计划退役

2026-08-31 的范围裁决已经由 #437 接管 taxonomy Gold admission 与 recorded Stable 分类测量，并明确禁止在真实结果出现前实施本 Issue。当前 37 项 taxonomy/TradeRelevance/policy 阈值、decide v12、holdout/shadow/canary/promotion 是基于未知结果提前设计的联合发布计划，不应继续作为 backlog 承诺。

#437 的结果不会自动授权 policy 变更。只有真实测量产生具体 failure cases 后，才新开一个只包含单一可解释 policy delta 的 calibration Issue。

Disposition: not planned。
